ogr/ogrsf_frmts/generic/ogrunionlayer.cpp

```diff
@@ -904,6 +904,108 @@ OGRErr OGRUnionLayer::IUpsertFeature(OGRFeature *poFeature)
     }
 }
 
+/************************************************************************/
+/*                           IUpdateFeature()                           */
+/************************************************************************/
+
+OGRErr OGRUnionLayer::IUpdateFeature(OGRFeature *poFeature,
+                                     int nUpdatedFieldsCount,
+                                     const int *panUpdatedFieldsIdx,
+                                     int nUpdatedGeomFieldsCount,
+                                     const int *panUpdatedGeomFieldsIdx,
+                                     bool bUpdateStyleString)
+{
+    if (!bPreserveSrcFID)
+    {
+        CPLError(CE_Failure, CPLE_NotSupported,
+                 "UpdateFeature() not supported when PreserveSrcFID is OFF");
+        return OGRERR_FAILURE;
+    }
+
+    if (osSourceLayerFieldName.empty())
+    {
+        CPLError(CE_Failure, CPLE_NotSupported,
+                 "UpdateFeature() not supported when SourceLayerFieldName is "
+                 "not set");
+        return OGRERR_FAILURE;
+    }
+
+    if (poFeature->GetFID() == OGRNullFID)
+    {
+        CPLError(CE_Failure, CPLE_NotSupported,
+                 "UpdateFeature() not supported when FID is not set");
+        return OGRERR_FAILURE;
+    }
+
+    if (!poFeature->IsFieldSetAndNotNull(0))
+    {
+        CPLError(CE_Failure, CPLE_NotSupported,
+                 "UpdateFeature() not supported when '%s' field is not set",
+                 osSourceLayerFieldName.c_str());
+        return OGRERR_FAILURE;
+    }
+
+    const char *pszSrcLayerName = poFeature->GetFieldAsString(0);
+    for (int i = 0; i < nSrcLayers; i++)
+    {
+        if (strcmp(pszSrcLayerName, papoSrcLayers[i]->GetName()) == 0)
+        {
+            pabModifiedLayers[i] = TRUE;
+
+            const auto poSrcLayerDefn = papoSrcLayers[i]->GetLayerDefn();
+            OGRFeature *poSrcFeature = new OGRFeature(poSrcLayerDefn);
+            poSrcFeature->SetFrom(poFeature, TRUE);
+            poSrcFeature->SetFID(poFeature->GetFID());
+
+            // We could potentially have a pre-computed map from indices in
+            // poLayerDefn to indices in poSrcLayerDefn
+            std::vector<int> anSrcUpdatedFieldIdx;
+            const auto poLayerDefn = GetLayerDefn();
+            for (int j = 0; j < nUpdatedFieldsCount; ++j)
+            {
+                if (panUpdatedFieldsIdx[j] != 0)
+                {
+                    const int nNewIdx = poSrcLayerDefn->GetFieldIndex(
+                        poLayerDefn->GetFieldDefn(panUpdatedFieldsIdx[j])
+                            ->GetNameRef());
+                    if (nNewIdx >= 0)
```

Comment:
Isn't the index check already done in  https://github.com/OSGeo/gdal/pull/7269/files#diff-a6dc28611df439c6421d715cce8094c14d0fd92c2402a907c170727ae9844993R739 ?

No objection for a second check here but maybe we should log the error.

Reply:
> Isn't the index check already done in

No, the check in L739 only checks the validity of values of panUpdatedFieldsIdx[], which are indices related to the *union* layer. Here we check that they map to a *source* layer field, as the union layer can be "larger" than its source layers. e.g if you have layer A with fieldA and layer B with fieldB, the union layer will expose fieldA and fieldB, and if you try to update fieldB of a feature belonging to layer A, then it dosn't map to anything. The question if that should be an error is interesting. Currently for ISetFeature() we also silently discards fields that don't map to the source layer definition. I'm not totally sure of the behaviour we want here. I'd keep as current for consistency with ISetFeature(). Anyway, I'm not sure update of union layers is used by anyone. This is mostly used internally by the OGRSQL engine to implement UNION ALL, and in read-only mode.